src/evm/protocol/vm/utils.rs

```diff
@@ -350,22 +355,161 @@ pub fn hexstring_to_vec(hexstring: &str) -> Result<Vec<u8>, SimulationError> {
+/// a `0x` prefix. It decodes the hex string into bytes, ensuring it does not exceed
+/// 32 bytes in length. If the string is valid and fits within 32 bytes, the bytes
+/// are copied into a `[u8; 32]` array, with right zero-padding for unused bytes.
+///
+/// # Arguments
+///
+/// * `pool_id` - A string slice representing a hexadecimal pool ID. It can optionally start with
+///   the `0x` prefix.
+///
+/// # Returns
+///
+/// * `Ok([u8; 32])` - On success, returns a 32-byte array with the decoded bytes. If the input is
+///   shorter than 32 bytes, the rest of the array is right padded with zeros.
+/// * `Err(SimulationError)` - Returns an error if:
+///     - The input string is not a valid hexadecimal string.
+///     - The decoded bytes exceed 32 bytes in length.
+///
+/// # Example
+/// ```
+/// use string_to_bytes32;
+///
+/// let pool_id = "0x1234abcd";
+/// match string_to_bytes32(pool_id) {
+///     Ok(bytes32) => println!("Bytes32: {:?}", bytes32),
+///     Err(e) => eprintln!("Error: {}", e),
+/// }
+/// ```
 pub fn string_to_bytes32(pool_id: &str) -> Result<[u8; 32], SimulationError> {
     let pool_id_no_prefix =
         if let Some(stripped) = pool_id.strip_prefix("0x") { stripped } else { pool_id };
     let bytes = hex::decode(pool_id_no_prefix)
         .map_err(|e| SimulationError::FatalError(format!("Invalid hex string: {}", e)))?;
-    if bytes.len() != 32 {
+    if bytes.len() > 32 {
         return Err(SimulationError::FatalError(format!(
-            "Hex string is not 32 bytes: length {}",
+            "Hex string exceeds 32 bytes: length {}",
             bytes.len()
         )));
     }
     let mut array = [0u8; 32];
-    array.copy_from_slice(&bytes);
+    array[..bytes.len()].copy_from_slice(&bytes);
     Ok(array)
 }
 
+/// Decodes a JSON-encoded list of hexadecimal strings into a `Vec<Vec<u8>>`.
+///
+/// This function parses a JSON array where each element is a string representing a hexadecimal
+/// value. It converts each hex string into a vector of bytes (`Vec<u8>`), and aggregates them into
+/// a `Vec<Vec<u8>>`.
+///
+/// # Arguments
+///
+/// * `input` - A byte slice (`&[u8]`) containing JSON-encoded data. The JSON must be a valid array
+///   of hex strings.
+///
+/// # Returns
+///
+/// * `Ok(Vec<Vec<u8>>)` - On success, returns a vector of byte vectors.
+/// * `Err(SimulationError)` - Returns an error if:
+///     - The input is not valid JSON.
+///     - The JSON is not an array.
+///     - Any array element is not a string.
+///     - Any string is not a valid hexadecimal string.
+///
+/// # Example
+/// ```
+/// use json_deserialize_address_list;
+///
+/// let json_input = br#"["0x1234", "0xc0ffee"]"#;
+/// match json_deserialize_address_list(json_input) {
+///     Ok(result) => println!("Decoded: {:?}", result),
+///     Err(e) => eprintln!("Error: {}", e),
+/// }
+/// ```
+pub fn json_deserialize_address_list(input: &[u8]) -> Result<Vec<Vec<u8>>, SimulationError> {
```

Comment:
where is this being used??

Reply:
It's the deserialize function related to [this serialize function](https://github.com/search?q=repo:propeller-heads/tycho-protocol-sdk%20json_serialize_address_list&type=code) that we have in the SDK. 

Currently it's only used by Balancer for rate providers.